projects/kubernetes-sigs/image-builder/patches/0028-Remove-old-kernels.patch

```diff
@@ -0,0 +1,90 @@
+From 6cb3dd7b4525eee17a935fca81d4ef0147bb58c4 Mon Sep 17 00:00:00 2001
+From: Vignesh Goutham Ganesh <vgg@amazon.com>
+Date: Thu, 14 Dec 2023 15:36:28 -0600
+Subject: [PATCH] Remove old kernels
+
+Signed-off-by: Vignesh Goutham Ganesh <vgg@amazon.com>
+---
+ eks-anywhere-checkout-v0.1.19                 |  0
+ eks-anywhere-patched                          |  0
+ images/capi/ansible/roles/node/tasks/main.yml |  3 ++
+ .../capi/ansible/roles/node/tasks/redhat.yml  | 36 +++++++++++++++++++
+ .../capi/ansible/roles/setup/tasks/redhat.yml |  2 +-
+ 5 files changed, 40 insertions(+), 1 deletion(-)
+ create mode 100644 eks-anywhere-checkout-v0.1.19
+ create mode 100644 eks-anywhere-patched
+ create mode 100644 images/capi/ansible/roles/node/tasks/redhat.yml
+
+diff --git a/images/capi/ansible/roles/node/tasks/main.yml b/images/capi/ansible/roles/node/tasks/main.yml
+index f9f6b9040..9120073b8 100644
+--- a/images/capi/ansible/roles/node/tasks/main.yml
++++ b/images/capi/ansible/roles/node/tasks/main.yml
+@@ -18,6 +18,9 @@
+ - import_tasks: amazonLinux2.yml
+   when: ansible_distribution == "Amazon"
+ 
++- import_tasks: redhat.yml
++  when: ansible_distribution == "RedHat"
++
+ # This is required until https://github.com/ansible/ansible/issues/77537 is fixed and used.
+ - name: Override Flatcar's OS family
+   set_fact:
+diff --git a/images/capi/ansible/roles/node/tasks/redhat.yml b/images/capi/ansible/roles/node/tasks/redhat.yml
+new file mode 100644
+index 000000000..c1a60150e
+--- /dev/null
++++ b/images/capi/ansible/roles/node/tasks/redhat.yml
+@@ -0,0 +1,36 @@
++# Copyright 2020 The Kubernetes Authors.
++
++# Licensed under the Apache License, Version 2.0 (the "License");
++# you may not use this file except in compliance with the License.
++# You may obtain a copy of the License at
++
++# http://www.apache.org/licenses/LICENSE-2.0
++
++# Unless required by applicable law or agreed to in writing, software
++# distributed under the License is distributed on an "AS IS" BASIS,
++# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
++# See the License for the specific language governing permissions and
++# limitations under the License.
++
++---
++- name: Get the list of installed kernels
++  ansible.builtin.shell: "rpm -q kernel --qf '%{version}-%{release}\n' | sort -V"
++  register: installed_kernels
++
++- name: Get the current running kernel
++  ansible.builtin.command: "uname -r"
++  register: current_kernel
++
++- name: Set the number of kernels to keep
++  set_fact:
++    kernels_to_keep: 1
++
++- name: Filter out the current kernels
++  set_fact:
++    old_kernels: "{{ installed_kernels.stdout_lines | difference([current_kernel.stdout]) | reverse | slice(kernels_to_keep)|list }}"
```

Comment:
Based on my understanding of this command, this is what's happening
For example:
`installed_kernels` = `[4.18, 5.10, 5.15]`
`current_kernel` = `5.15`
`kernels_to_keep` = `1`
`installed_kernels.stdout_lines | difference([current_kernel.stdout])` = `[4.18, 5.10]`
`... | reverse` = `[5.10, 4.18]`
`... | slice(kernels_to_keep) | list` = `[5.10]` ===> `old_kernels` = `[5.10]`

if my understanding is correct, aren't we skipping removing the 4.18 kernel in the above case, or are we just making the assumption that exactly one old kernel will be there besides the current one? Or am I interpreting it incorrectly?

Reply:
Yes, there will always be only 1 extra kernel present on the image. But I missed adding my changes, this command is not needed to be this extensive. My test OS had 3 kernels for some reason which is not the case with base ISOs. Updating PR in a bit.